### ISSUE TITLE
Deadlock with parallel processing of single partition stream

### ISSUE THREAD

**Issue by vladimirkl** [zio/zio-kafka#852]
Hi, I need to take n records from single partition, process in parallel batches, and commit. Unfortunately this code causes a deadlock in Consumer on commit:

```scala
 val consumerLayer: ZLayer[Kafka, Throwable, Consumer] =
     ZLayer.scoped ( ZIO.serviceWithZIO[Kafka] { kafka =>
       Consumer.make(ConsumerSettings(kafka.bootstrapServers).withGroupId("group1").withOffsetRetrieval(Consumer.OffsetRetrieval.Auto(AutoOffsetStrategy.Earliest)))
     })

...

Producer.produceChunk(Chunk.fromIterable(1 to 1000).map(n => new ProducerRecord(topic, n, n.toString)), Serde.int, Serde.string) *>
  Consumer.plainStream(Subscription.topics(topic), Serde.int, Serde.string)
    .take(100)
    .groupedWithin(10, 100.millis)
    .mapZIOPar(2)(c => ZIO.debug(c.size) as c.map(_.offset))
    .map(OffsetBatch.apply)
    .debug("Offset")
    .mapZIO(_.commit)
    .debug("Commit")
    .runDrain zipPar Fiber.dumpAll.delay(20.seconds)

```
Replacing `groupedWithin` with `grouped` makes this code working. Also I can leave `groupedWithin` in place and remove `take` -  no deadlock occurs but I need to terminate stream exactly for n records.  It looks like race condition on stream termination with `take`.  Tested with zio-kafka 2.3.0, Scala 2.13.10  and embedded Kafka

**Comment by guizmaii** [comment]
Thanks for your report @vladimirkl 🙂

I'm wondering if this is a bug related to zio-kafka or zio-streams 🤔
Because, for example, changing `groupedWithin` with `grouped` as you did, has nothing to do with zio-kafka.
Why do you think it comes from zio-kafka?

Sorry for the questions. I'm trying to better understand your issue.

**Comment by vladimirkl** [comment]
It may be some zio-streams issue, but I never encountered it without zio-kafka - while I have similar aggregations in other places . It locks exactly on `commit`. If you wait some time after deadlock you will see RunloopTimeout exception. Without `commit` take terminates stream correctly

**Comment by vladimirkl** [comment]
I even tried to replace `commit` with something non interruptible  like `ZIO.attemptBlocking(Thread.sleep(2000))` and it works correctly. So it's very likely  Consumer issue

**Comment by guizmaii** [comment]
Thanks for the additional details :)

**Comment by erikvanoosten** [comment]
I used this program to test the issue: https://gist.github.com/erikvanoosten/5e9f34d8ff43de32b583c021c858e309

This problem is caused by an immediate unsubscribe after the stream ends (see https://github.com/zio/zio-kafka/blob/master/zio-kafka/src/main/scala/zio/kafka/consumer/Consumer.scala#L254). Since we are no longer subscribed, we (can) no longer poll. When we stop polling, the commit callbacks will not be called anymore and all progress is halted.

Potential solutions:

* delay unsubscribe until there are no more pending commits
* upon unsubscribe, cancel all pending commits

@svroonland WDYT?

**Comment by guizmaii** [comment]
@erikvanoosten What is calling the `unsubscribe` in @vladimirkl example? Just the end of the stream?

**Comment by vladimirkl** [comment]
Why same code works with `grouped` instead of `groupedWithin`? Perhaps order of finalization is different?

**Comment by erikvanoosten** [comment]
> Why same code works with `grouped` instead of `groupedWithin`? Perhaps order of finalization is different?

`groupedWithin` is time based. This causes a fiber split, everything above the groupedWithin runs on another fiber than what follows. I suspect this causes the unsubscribe to happen fractionally later than with `grouped`.

**Comment by vladimirkl** [comment]
Things are much worse with zio-streams 2.0.14. Even simpler example hangs forever:
```scala
    Consumer.plainStream(Subscription.topics(topic), Serde.int, Serde.string)
      .take(100)
      .map(_.offset)
      .aggregateAsync(Consumer.offsetBatches)
      .debug("Offset")
      .mapZIO(_.commit)
      .debug("Commit")
      .runDrain
```
This code works perfectly with zio-streams 2.0.13. Unfortunately `take` becomes dangerous for `zio-kafka` streams at all.

**Comment by svroonland** [comment]
@erikvanoosten I believe your analysis is correct, the `take` ends the stream and unsubscribes. I have similar issues with `take` and finalizer race conditions in zio-kinesis. 

Regarding the possible solutions: if it's a race condition, then I'm not sure we always have pending commits to await before unsubscribing.

We could look into a usage pattern that uses graceful shutdown to end the stream but keep the subscription.

**Comment by guizmaii** [comment]
@svroonland Look at #890. The issue seems to be with the `commitAsync`

**Comment by svroonland** [comment]
As in, commitAsync requires poll calls to complete? Yeah

**Comment by vladimirkl** [comment]
I encountered few more issues with hanging commit in other scenarios - when broker dies, then dies KafkaConsumer, but async commit hangs. Compared with fs2-kafka implementation - it has similar behaviour, but uses timeout for commit operation (15 seconds by default). zip-kafka user can add timeouts to commit everywhere in their code, but I think it's a good idea to add default timeout for safety - similar to fs2-kafka.  I can create a PR. What do you think?

**Comment by guizmaii** [comment]
>  I can create a PR. What do you think?

Please do :)

**Comment by guizmaii** [comment]
Isn't this issue fixed by https://github.com/zio/zio-kafka/pull/982? Can we close it?

**Comment by vladimirkl** [comment]
Not sure - timeout is definitely better than deadlock, but original issue with async grouping still remains. However if we cannot handle it all, we can close it for now.

**Comment by erikvanoosten** [comment]
Fixed in #1123.
